src/main/kotlin/org/rust/ide/intentions/ConvertClosureToFunctionIntention.kt

```diff
@@ -0,0 +1,135 @@
+
+    data class Context(
+        val assignment: RsLetDecl,
+        val lambda: RsLambdaExpr,
+    )
+
+    private fun getLetDeclarationForElementInSignature(element: PsiElement): RsLetDecl? {
+        if (element.text == ";") {
+            return null
+        }
+
+        for (el in element.ancestors) {
+            return when (el) {
+                is RsLetDecl -> el
+                is RsValueArgumentList -> el.ancestorStrict()
+                is RsBlock -> return null
+                else -> continue
+            }
+        }
+
+        return null
+    }
+
+    override fun findApplicableContext(project: Project, editor: Editor, element: PsiElement): Context? {
+        // We try to find a let declaration
+        val possibleTarget = element.ancestorStrict<RsLetDecl>() ?: return null
+
+        // The assignment of the let declaration should be a lambda to be a valid target
+        if (possibleTarget.expr !is RsLambdaExpr) {
+            return null
+        }
+
+        val signatureTarget = getLetDeclarationForElementInSignature(element) ?: return null
+
+        return Context(signatureTarget, signatureTarget.expr as RsLambdaExpr)
+    }
+
+    override fun invoke(project: Project, editor: Editor, ctx: Context) {
+        val factory = RsPsiFactory(project)
+
+        var targetName = ctx.assignment.pat?.text
+
+        var isTemplate = if (targetName == null || !isValidRustVariableIdentifier(targetName)) {
+            targetName = "func"
+
+            true
+        } else {
+            false
+        }
+
+        val valueParameters = ctx.lambda.valueParameters
+        val (parametersText, parameterPlaceholders) = this.createParameterList(factory, valueParameters)
+        if (parameterPlaceholders.isNotEmpty()) {
+            isTemplate = true
+        }
+        val lambdaExpr = ctx.lambda.expr
+        val returnText = if (ctx.lambda.retType != null) {
+            ctx.lambda.retType!!.text
+        } else if (lambdaExpr != null && lambdaExpr.type != TyUnknown && lambdaExpr.type != TyUnit){
+            "-> ${lambdaExpr.type.render()}"
+        } else {
+            ""
+        }
+
+        val body = if (lambdaExpr is RsBlockExpr) {
+            lambdaExpr.text
+        } else {
+            "{ ${lambdaExpr?.text} }"
+        }
+
+        val function = factory.createFunction("fn $targetName($parametersText) $returnText $body")
+        val replaced = ctx.assignment.replace(function) as RsFunction
+
+        val identifier = replaced.identifier
+
+        // in case we auto-generated a function name, we encourage the user to change it by running a template on the replacement
+        if (isTemplate) {
+            val placeholderElements = mutableListOf<SmartPsiElementPointer<PsiElement>>()
+            if (targetName == "func") {
+                placeholderElements.add(identifier.createSmartPointer())
+            }
+            placeholderElements.addAll(replaced.descendantsOfType<RsTypeReference>().filter {
```

Comment:
This would also use the template i.e. for unit type reference inside the generated function body.
What about returning a list of parameter indices that need to be replaced from `createParameterList` and then using this index list to select the correct parameters to put into the template builder?